base/abstractarray.jl

```diff
@@ -2399,3 +2399,11 @@ function hash(A::AbstractArray, h::UInt)
 
     return h
 end
+
+# The semantics of `collect` are weird. Better to write our own
```

Comment:
How about `collect(T, Iterators.rest(a, state))`? Then for the 1-argument case `Vector{T}(a)`. Better to avoid repeated `push!` since it's a bit slow.

Reply:
The problem with that is this:
```
julia> collect(Int, Iterators.rest([1 2; 3 4]))
2×2 Array{Int64,2}:
 1  2
 3  4

julia> collect(Int, Iterators.rest([1 2; 3 4], 1))
4-element Array{Int64,1}:
 1
 3
 2
 4
```
Perhaps we can solve that with doing either a `reshape` or just a single-slice `view` of `a` first though.

Reply:
> Then for the 1-argument case Vector{T}(a).

That won't work for the multi-dimensional case, unfortunately:
```
julia> Vector{Int}([1 2; 3 4])
ERROR: MethodError: no method matching Array{Int64,1}(::Array{Int64,2})
Closest candidates are:
  Array{Int64,1}(::AbstractArray{S,N}) where {T, N, S} at array.jl:562
  Array{Int64,1}() where T at boot.jl:425
  Array{Int64,1}(::UndefInitializer, ::Int64) where T at boot.jl:406
  ...
Stacktrace:
 [1] top-level scope at REPL[25]:1
```
`collect` also doesn't always return a `Vector` and I think it would be weird if `Base.rest(::A)` and `Base.rest(::A, st)` returned different types.

Reply:
Also, note that:
```
julia> Base.IteratorSize(Iterators.rest([1 2; 3 4], 1))
Base.SizeUnknown()
```
so `collect` will end up just using `push!` anyways.